src/Compilers/Core/MSBuildTask/ArgsTaskItem.cs

```diff
@@ -0,0 +1,98 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using Microsoft.Build.Framework;
+
+namespace Microsoft.CodeAnalysis.BuildTasks
+{
+    internal sealed class ArgsTaskItem : ITaskItem
+    {
+        // This list is taken from https://github.com/dotnet/msbuild/blob/291a8108761ed347562228f2f8f25477996a5a93/src/Shared/Modifiers.cs#L36-L70
+        private static readonly string[] WellKnownItemSpecMetadataNames =
+        [
+            "FullPath",
+            "RootDir",
+            "Filename",
+            "Extension",
+            "RelativeDir",
+            "Directory",
+            "RecursiveDir",
+            "Identity",
+            "ModifiedTime",
+            "CreatedTime",
+            "AccessedTime",
+            "DefiningProjectFullPath",
+            "DefiningProjectDirectory",
+            "DefiningProjectName",
+            "DefiningProjectExtension",
+        ];
+
+        private readonly Dictionary<string, string> _metadata = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+
+        public ArgsTaskItem(string itemSpec)
+        {
+            ItemSpec = itemSpec;
+        }
+
+        public string ItemSpec { get; set; }
+
+        // Implementation notes that we should include the built-in metadata names as well as our custom ones.
+        public ICollection MetadataNames
+        {
+            get
+            {
+                var clone = new List<string>(_metadata.Keys);
+                clone.AddRange(WellKnownItemSpecMetadataNames);
+                return clone;
+            }
+        }
+
+        // Implementation notes that we should include the built-in metadata names as well as our custom ones.
+        public int MetadataCount => _metadata.Count + WellKnownItemSpecMetadataNames.Length;
+
+        public IDictionary CloneCustomMetadata()
+        {
+            return new Dictionary<string, string>(_metadata, StringComparer.OrdinalIgnoreCase);
+        }
+
+        public void CopyMetadataTo(ITaskItem destinationItem)
+        {
+            // Implementation notes that we should not overwrite existing metadata on the destination.
+            var destinationMetadataNames = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+            foreach (var name in destinationItem.MetadataNames)
+            {
+                destinationMetadataNames.Add((string)name);
+            }
+
+            foreach (var metadataName in _metadata.Keys)
+            {
+                if (destinationMetadataNames.Contains(metadataName))
+                {
+                    continue;
+                }
+
+                var metadataValue = _metadata[metadataName];
+                destinationItem.SetMetadata(metadataName, metadataValue);
+            }
+        }
+
+        public string GetMetadata(string metadataName)
```

Comment:
@rainersigwald, @baronfel can you help me understand the contract of `ITaskItem` a bit better? One part of this impl, and others I've seen, is that they have this pattern of exporting a set of names via `MetadataNames` that will throw when you pass them to `GetMetadata`. My mind sees that as a dictionary like type exporting a key but then throwing when you ask for the value. Why does `ITaskItem` behave this way?

Reply:
Do you have a specific instance we could look at? From what we can see TaskItem MetadataNames + GetMetadata _should_ be safe for all of the returned names, which include explicit Item Metadata as well as the historically-significant set of metadata that all Items have (because all items are files, don'tchanknow).

Reply:
Maybe I misinterpreted the [comments on the interface](https://github.com/dotnet/msbuild/blob/291a8108761ed347562228f2f8f25477996a5a93/src/Framework/ITaskItem.cs#L35) and the [implementation here](https://github.com/dotnet/msbuild/blob/291a8108761ed347562228f2f8f25477996a5a93/src/Utilities/TaskItem.cs#L223). `MetadataNames` includes the "built-in" metadata which I wasn't sure would actually by in my custom metadata dictionary.